### PR TITLE
Fixes after Nomic Labs audit

### LINKED ISSUE #70: Improve memory handling of the Proxy contract
Some issues here:
- the free memory pointer is never updated
- we are using `ptr` again [here](https://github.com/zeppelinos/zos-lib/blob/40e09a31c9f1096baaff5f6b84d7c56a9600c871/contracts/upgradeability/Proxy.sol#L25) without re-reading it.

### LINKED ISSUE #75: Optimize slot number hash calculation
Currently the hashes we use for proxy slots are recalculated in every transaction.

https://github.com/zeppelinos/zos-lib/blob/2cf0aad3e4fb4d97e694b74fa73e9de680214657/contracts/upgradeability/UpgradeabilityProxy.sol#L18

We should write the precalculated values and then in the constructor check them against a hash obtained on chain.

### PR DESCRIPTION
**Pull request by frangio**
This PR fixes most of the issues reported in the [Nomic Labs audit](https://medium.com/nomic-labs-blog/zeppelinos-smart-contracts-audit-c2173c7753dd), and explains the issues we have decided not to fix.

Issues:
- [x] `UpgradeabilityProxy` is vulnerable to function selector clashing (fixed in #36)
- [x] `ContractDirectory#setImplementation` doesn't check if the implementation is a contract

Recommendations:
- [x] `Proxy` fallback function should be `external` (fixed in https://github.com/zeppelinos/zos-lib/pull/36)
- [x] No need to use free memory pointer in `Proxy` fallback (closes #70)
- [x] Hash of storage slot keys is calculated in every run (closes #75)
- [x] Explicitly cast contract to address for `.call` operations
- [x] Rename `setUpgradeabilityOwner` to `setProxyOwner` for consistency (fixed in #36)
- [x] Make `Package#getVersion` consistent  with all the other getters (return 0 rather than revert on error)
- [x] Add `Initializable` tests (fixed in #82)

---

Not fixed:

- [ ] **Migrations may be applied in arbitrary order**
We are planning to change the design for migrations soon after the MVP release. For now, we have added warnings to the documentation for `Migratable` and `Initializable`.

- [ ] **`Initializable` contracts can't be inherited multiple times**
Likewise, we are planning to change the design for initializers soon. In the meantime we are using `Migratable` to mitigate this issue.

- [ ] **`BaseAppManager#upgradeTo[AndCall]` can be called with an incorrect contract name**
This functionality is exposed through the CLI in a way which makes such an error impossible. However, we have some ideas on how to make this a safer API post-MVP. In the meantime we are leaving the functions as-is because at this stage we prefer to give flexibility to the users.

- [ ] **Abstract inline assembly of `[Owned]UpgradeabilityStorage` into functions**
We wouldn't be able to give a type to such a function because the return value is unknown. We prefer to leave the small amount of ad-hoc inline assembly.

- [ ] **Consider using `pragma experimental "v0.5.0"`, which is recommended since 0.4.21**
Will consider after MVP release.

- [ ] **Add tests for different types of upgrades**
We've agreed with Nomic that the current upgradeability implementation is sound. I've created https://github.com/zeppelinos/zos-lib/issues/117 to enhance the tests later.

### CODE REVIEW

**Comment by facuspagnuolo** [comment]
@frangio please update with master

**Comment by frangio** [comment]
@facuspagnuolo Rebased and fixed the new tests that were failing. :slightly_smiling_face:

**Comment by facuspagnuolo** [comment]
@frangio `Add Initializable tests` could be considered done, right?

File: contracts/application/versioning/ImplementationDirectory.sol [hunk 1]
```diff
@@ -33,7 +34,12 @@ contract ImplementationDirectory is ImplementationProvider, Ownable {
    * @param implementation Address of the implementation
    */
   function setImplementation(string contractName, address implementation) public onlyOwner {
+    require(AddressUtils.isContract(implementation));
     implementations[contractName] = implementation;
     emit ImplementationChanged(contractName, implementation);
   }
+
+  function unsetImplementation(string contractName) public onlyOwner {
+    implementations[contractName] = address(0);
```
Comment (facuspagnuolo) [inline contracts/application/versioning/ImplementationDirectory.sol]: Shouldn't we emit an event here to be consistent?

File: contracts/upgradeability/AdminUpgradeabilityProxy.sol [hunk 2]
```diff
@@ -21,7 +21,8 @@ contract AdminUpgradeabilityProxy is UpgradeabilityProxy {
   event AdminChanged(address previousAdmin, address newAdmin);
 
   // Storage slot of the admin of the contract
-  bytes32 private constant ADMIN_SLOT = keccak256("org.zeppelinos.proxy.admin");
+  bytes32 private constant ADMIN_SLOT =
+    0x10d6a54a4754c8869d6886b5f5d7fbfa5b4522237ea5c60d11bc4e7a1ff9390b;
```
Comment (facuspagnuolo) [inline contracts/upgradeability/AdminUpgradeabilityProxy.sol]: Nit: it's weird to have multiple-line assignments.
I'd a comment here saying this value is `keccak256("org.zeppelinos.proxy.admin")` (I know we are asserting that in the `constructor`)

File: contracts/upgradeability/Proxy.sol [hunk 3]
```diff
@@ -32,20 +32,22 @@ contract Proxy {
...
       // Copy the returned data.
-      returndatacopy(ptr, 0, returndatasize)
+      returndatacopy(0, 0, returndatasize)
 
       switch result
       // delegatecall returns 0 on error.
-      case 0 { revert(ptr, returndatasize) }
-      default { return(ptr, returndatasize) }
+      case 0 { revert(0, returndatasize) }
+      default { return(0, returndatasize) }
```
Comment (facuspagnuolo) [inline contracts/upgradeability/Proxy.sol]: I'm not totally fine with these changes, will discuss them offline

File: contracts/upgradeability/UpgradeabilityProxy.sol [hunk 4]
```diff
@@ -18,12 +18,15 @@ contract UpgradeabilityProxy is Proxy {
   event Upgraded(address implementation);
 
   // Storage slot of the address of the current implementation
-  bytes32 private constant IMPLEMENTATION_SLOT = keccak256("org.zeppelinos.proxy.implementation");
+  bytes32 private constant IMPLEMENTATION_SLOT  =
+    0x7050c9e0f4ca769c69bd3a8ef740bc37934f8e2c036e5a723fd8ee048ed3f8c3;
```
Comment (facuspagnuolo) [inline contracts/upgradeability/UpgradeabilityProxy.sol]: same above

File: test/application/versioning/Release.test.js [hunk 5]
```diff
@@ -52,11 +53,14 @@ contract('Release', ([_, developer, anotherAddress, implementation_v0, implement
   })
 
   describe('setImplementation', function () {
+    beforeEach(async function() {
+      this.implementation_v0 = (await DummyImplementation.new()).address
+      this.implementation_v1 = (await DummyImplementation.new()).address
```
Comment (facuspagnuolo) [inline test/application/versioning/Release.test.js]: I think you are not using `this.implementation_v0`. Additionally, I'd remove the `implementation_v0` and `implementation_v1` variables declared above in the `contracts` context block.

**Review by facuspagnuolo — COMMENTED** [review]
Looking good, left some comments

File: test/application/versioning/Release.test.js (hunk 5 above)
Comment (frangio) [reply test/application/versioning/Release.test.js]: `this.implementation_v0` is used in `shouldBehaveLikeImplementationDirectory`. It isn't used in the `when it is frozen` but I left it for simplicity.

**Review by facuspagnuolo — APPROVED** [review]
Great work @frangio ! 👏

File: contracts/upgradeability/UpgradeabilityProxy.sol [hunk 6]
```diff
@@ -17,13 +17,19 @@ contract UpgradeabilityProxy is Proxy {
...
   event Upgraded(address implementation);
 
-  // Storage slot of the address of the current implementation
-  bytes32 private constant IMPLEMENTATION_SLOT = keccak256("org.zeppelinos.proxy.implementation");
+  /**
+   * @dev Storage slot with the address of the current implementation.
+   * @dev This is the hash of "org.zeppelinos.proxy.implementation", and is
+   * @dev validated in the constructor.
+   */
+  bytes32 private constant IMPLEMENTATION_SLOT = 0x7050c9e0f4ca769c69bd3a8ef740bc37934f8e2c036e5a723fd8ee048ed3f8c3;
```
Comment (k06a) [inline contracts/upgradeability/UpgradeabilityProxy.sol:25]: Why this change was made?

File: contracts/upgradeability/AdminUpgradeabilityProxy.sol (hunk 2 above)
Comment (k06a) [reply contracts/upgradeability/AdminUpgradeabilityProxy.sol]: @facuspagnuolo but why these changes were made?

Comment (k06a) [reply contracts/upgradeability/AdminUpgradeabilityProxy.sol]: Just got it: https://github.com/zeppelinos/zos-lib/issues/75

File: contracts/upgradeability/UpgradeabilityProxy.sol (hunk 6 above)
Comment (frangio) [reply contracts/upgradeability/UpgradeabilityProxy.sol:25]: Because Solidity inlined a `keccak256` calculation everywhere the variable was used, instead of calculating the value during compilation. This resulted in a lot of additional cost for every interaction with the proxy.
